File: src/main/java/xyz/upperlevel/hgame/network/NettyChannelInitializer.java
package xyz.upperlevel.hgame.network;

import io.netty.channel.ChannelInitializer;
import io.netty.channel.ChannelPipeline;
import io.netty.channel.socket.SocketChannel;
import io.netty.handler.codec.LengthFieldBasedFrameDecoder;
import io.netty.handler.codec.LengthFieldPrepender;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class NettyChannelInitializer extends ChannelInitializer<SocketChannel> {
    private final Endpoint endpoint;

    @Override
    protected void initChannel(SocketChannel ch) throws Exception {
        ChannelPipeline p = ch.pipeline();
        p.addLast("frameDecoder", new LengthFieldBasedFrameDecoder(Integer.MAX_VALUE, 0, 4, 0, 4));
        p.addLast("frameEncoder", new LengthFieldPrepender(4));
        p.addLast("codec", new PayloadPacketCodec(endpoint.getJsonMapper()));
        p.addLast("executor", new NettyChannelEventCaller(endpoint));
    }
}

File: game/src/main/kotlin/xyz/upperlevel/hgame/network/NettyChannelInitializer.kt
package xyz.upperlevel.hgame.network

import io.netty.channel.ChannelInitializer
import io.netty.channel.socket.SocketChannel
import io.netty.handler.codec.LengthFieldBasedFrameDecoder
import io.netty.handler.codec.LengthFieldPrepender


class NettyChannelInitializer(private val endpoint: Endpoint) : ChannelInitializer<SocketChannel>() {
    override fun initChannel(ch: SocketChannel) {
        val p = ch.pipeline()
        p.addLast("frameDecoder", LengthFieldBasedFrameDecoder(Integer.MAX_VALUE, 0, 4, 0, 4))
        p.addLast("frameEncoder", LengthFieldPrepender(4))
        p.addLast("codec", PayloadPacketCodec(endpoint.jsonMapper))
        p.addLast("executor", NettyChannelEventCaller(endpoint))
    }
}
